test/backend/x11/test_window.py

```diff
@@ -0,0 +1,744 @@
+        assert xmanager.c.window.info()['width'] == 618
+        assert xmanager.c.window.info()['height'] == 468
+        assert xmanager.c.window.info()['x'] == 820
+        assert xmanager.c.window.info()['y'] == 0
+    finally:
+        for win in w:
+            win.kill_client()
+        conn.finalize()
+
+    test_initial_state()
+
+
+class BringFrontClickConfig(ManagerConfig):
+    bring_front_click = True
+
+
+class BringFrontClickFloatingOnlyConfig(ManagerConfig):
+    bring_front_click = "floating_only"
+
+
+@pytest.fixture
+def bring_front_click(request):
+    return request.param
+
+
+@pytest.mark.parametrize(
+    "xmanager, bring_front_click",
+    [
+        (ManagerConfig, False),
+        (BringFrontClickConfig, True),
+        (BringFrontClickFloatingOnlyConfig, "floating_only"),
+    ],
+    indirect=True,
+)
+def test_bring_front_click(xmanager, bring_front_click):
+    def get_all_windows(conn):
+        root = conn.default_screen.root.wid
+        q = conn.conn.core.QueryTree(root).reply()
+        return list(q.children)
+
+    def fake_click(conn, xtest, x, y):
+        root = conn.default_screen.root.wid
+        xtest.FakeInput(6, 0, xcffib.xproto.Time.CurrentTime, root, x, y, 0)
+        xtest.FakeInput(4, 1, xcffib.xproto.Time.CurrentTime, root, 0, 0, 0)
+        xtest.FakeInput(5, 1, xcffib.xproto.Time.CurrentTime, root, 0, 0, 0)
+        conn.conn.flush()
+
+    conn = xcbq.Connection(xmanager.display)
+    xtest = conn.conn(xcffib.xtest.key)
+
+    # this is a tiled window.
+    xmanager.test_window("one")
+    xmanager.c.sync()
+
+    xmanager.test_window("two")
+    xmanager.c.window.set_position_floating(50, 50)
+    xmanager.c.window.set_size_floating(50, 50)
+    xmanager.c.sync()
+
+    xmanager.test_window("three")
+    xmanager.c.window.set_position_floating(150, 50)
+    xmanager.c.window.set_size_floating(50, 50)
+    xmanager.c.sync()
+
+    wids = [x["id"] for x in xmanager.c.windows()]
+    names = [x["name"] for x in xmanager.c.windows()]
+
+    assert names == ["one", "two", "three"]
+    wins = get_all_windows(conn)
+    assert wins.index(wids[0]) < wins.index(wids[1]) < wins.index(wids[2])
+
+    # Click on window two
+    fake_click(conn, xtest, 55, 55)
+    xmanager.c.sync()
+    wins = get_all_windows(conn)
+    if bring_front_click:
+        assert wins.index(wids[0]) < wins.index(wids[2]) < wins.index(wids[1])
+    else:
+        assert wins.index(wids[0]) < wins.index(wids[1]) < wins.index(wids[2])
+
+    # Click on window one
+    fake_click(conn, xtest, 10, 10)
+    xmanager.c.sync()
+    wins = get_all_windows(conn)
+    if bring_front_click == "floating_only":
+        assert wins.index(wids[0]) < wins.index(wids[2]) < wins.index(wids[1])
+    elif bring_front_click:
+        assert wins.index(wids[2]) < wins.index(wids[1]) < wins.index(wids[0])
+    else:
+        assert wins.index(wids[0]) < wins.index(wids[1]) < wins.index(wids[2])
```

Comment:
This function (and a lot of others too) don't feel X11-specific to me, even if the implementation is quite tied to X11 currently. :/

Reply:
I see what you mean - the behaviour itself is not specific to the X11 _backend_, but the implementation of Z stacking (in this case) and therefore our testing tools are specific to the environment. In the Wayland equivalent, much of the code will be similar, but our determination of whether the outcomes are correct will be specific to that environment.

Reply:
What testing tools? `get_all_windows` and `fake_click`? `get_all_windows` could be a function inside the manager and `fake_click` a backend-agnostic function in tests helpers.

Reply:
ooh nice, i like it. Will give that a shot.

Reply:
I set these two functions and managed to move `test_popup` in addition to this `bring_front_click` test back into the main set of tests for use with both backends. Works great! I think the `cursor_warp` tests might also be able to be re-used but I'm less certain there so I've left it until there is actually more setup on the Wayland test side.